### PR TITLE
Support ping protocol

### PR DESCRIPTION
**Pull request by jjyr**

### CODE REVIEW

File: ping/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
+mod protocol_builder;
+mod protocol_generated;
```
Comment (driftluo) [inline ping/src/lib.rs]: use `#[rustfmt::skip]`/`#[allow(clippy::all)]` on here, skip flatbuffer generate code

File: ping/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    UnexpectedError(PeerId),
+}
+
+pub struct PingProtocol {
+    id: ProtocolId,
+    /// the interval that we send ping to peers.
+    interval: Duration,
+    /// consider peer is timeout if during a timeout we still have not received pong from a peer
+    timeout: Duration,
+    event_sender: Sender<Event>,
```
Comment (driftluo) [inline ping/src/lib.rs]: Depending on the context, and the example, I think there might not need a `futures sender` here(maybe a crossbeam channel?), or consider implementing a `newtype` structure?

File: ping/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+use std::{
+    str,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tokio::codec::length_delimited::LengthDelimitedCodec;
+
+const SEND_PING_TOKEN: u64 = 0;
+const CHECK_TIMEOUT_TOKEN: u64 = 1;
+// TODO: replace this with real PeerId
+type PeerId = SessionId;
```
Comment (driftluo) [inline ping/src/lib.rs]: `peerid` just a hash for public key，I think this is a very easy thing to implement, even in the upper layer agreement. But I am not sure what Hash algorithm to use.

File: ping/src/lib.rs (hunk 2 above)
Comment (jjyr) [reply ping/src/lib.rs]: Why?

File: ping/src/lib.rs (hunk 3 above)
Comment (jjyr) [reply ping/src/lib.rs]: In our context, we need a hash function that security enough(avoid collision between different public keys) and should be as fast as possible.

https://dadario.com.br/cryptographic-and-non-cryptographic-hash-functions/
https://github.com/oconnor663/blake2b_simd

IMO blake2b is a good choice, it has the same security level with sha3, and benchmark shows it also has high performance. (In our scenario we need to hash a public key(65 bytes) and the performance result may be different, we need to do some benchmark to make sure).

I think we cloud use blake2b, we can replace it anytime before main-net launch.

**Comment by driftluo** [comment]
Wait to upgrade `generic-channel` to 0.2

**Pull request merged**